content_scripts/VirtualCamera.js

```diff
@@ -402,22 +513,27 @@ import * as THREE from '../../thirdPartyCode/three/three.module.js';
             const LOWER_ANGLE = Math.PI * 0.2; // soft lower bound
             if (verticalAngle > LOWER_ANGLE && verticalAngle < UPPER_ANGLE) {
                 // if within soft bounds, move the camera as usual
-                this.position = add(this.position, this.velocity);
-                this.targetPosition = add(this.targetPosition, this.targetVelocity);
+                if (this.mouseInput.isRotateRequested) {
+                    this.position = addButKeepDistance(this.position, this.velocity, this.mouseInput.lastWorldPos);
+                    this.targetPosition = addButKeepDistance(this.targetPosition, this.targetVelocity, this.mouseInput.lastWorldPos);
+                } else {
+                    this.position = add(this.position, this.velocity);
+                    this.targetPosition = add(this.targetPosition, this.targetVelocity);
+                }
             } else {
                 // if between soft bounds and top or bottom, slow movement exponentially as angle approaches 0 or PI
                 // e.g. if angle is PI * 0.85, we are 25% between UPPER_ANGLE and PI, so slow down by 0.25^2 = 6%
                 // e.g. if angle is PI * 0.9, we are 50% between UPPER_ANGLE and PI, so slow down by 0.50^2 = 25%
                 // e.g. if angle is PI * 0.99, we are 95% between UPPER_ANGLE and PI, so slow down by 0.95^2 = 90%
                 let closenessToAbsoluteBorder = (verticalAngle > Math.PI / 2) ? ((verticalAngle - UPPER_ANGLE) / (Math.PI - UPPER_ANGLE)) : ((verticalAngle - LOWER_ANGLE) / (0 - LOWER_ANGLE));
                 let scaleFactor = Math.pow(1 - closenessToAbsoluteBorder, 2);
-                this.position = add(this.position, scalarMultiply(this.velocity, scaleFactor));
-                this.targetPosition = add(this.targetPosition, scalarMultiply(this.targetVelocity, scaleFactor));
-            }
-
-            // if rotating, and distance has drifted without intentionally zooming, reset back to correct distance
-            if (this.preRotateDistanceToTarget && Math.abs(this.preRotateDistanceToTarget - this.distanceToTarget) > 10) {
-                this.zoomBackToPreRotateLevel();
+                if (this.mouseInput.isRotateRequested) {
+                    this.position = addButKeepDistance(this.position, scalarMultiply(this.velocity, scaleFactor), this.mouseInput.lastWorldPos);
+                    this.targetPosition = addButKeepDistance(this.targetPosition, scalarMultiply(this.targetVelocity, scaleFactor), this.mouseInput.lastWorldPos);
+                } else {
+                    this.position = add(this.position, scalarMultiply(this.velocity, scaleFactor));
+                    this.targetPosition = add(this.targetPosition, scalarMultiply(this.targetVelocity, scaleFactor));
+                }
             }
```

Comment:
I think this needs another look, to implement some sort of clamped vertical angle that you can rotate to, otherwise the camera flickers when it hits the top.

In my previous code, I was multiplying all rotate and strafe velocities by a smaller and smaller number the closer you approach the top or bottom angle. This prevented you from reaching the "singularity" point where it flickers. As far as I can tell, your version is somehow allowing rotate to happen with full velocity even when it's close to the top. But it's slowing down the strafe. I realized we don't actually need to multiply the strafe velocity by `scaleFactor`, so I think you can remove that. But I think you need some new logic to clamp the vertical angle, as the previous version doesn't seem to work anymore.

Reply:
Sounds good. I'm working on clamping the vertical angles. Will do a PR when it's done.